<div style="background: linear-gradient(135deg, #1D428A 0%, #C8102E 50%, #1D428A 100%); padding: 40px 30px; border-radius: 15px; margin-bottom: 20px;">
    <h1 style="color: #FFFFFF; margin: 0; font-size: 42px; text-align: center;">🏀 NBA Player Archetypes</h1>
    <h3 style="color: #C0C0C0; text-align: center; font-weight: 300; margin-top: 10px;">How the NBA's Player Roles Have Evolved (1996–2022)</h3>
    <p style="color: #A0A0A0; text-align: center; font-size: 14px; margin-top: 15px;">K-Means Clustering • PCA Visualization • Positional Entropy • Unicorn Index • Efficiency Trends</p>
</div>

---

## 📊 Executive Summary

<div style="background-color: #f0f4f8; padding: 20px; border-radius: 10px; border-left: 5px solid #1D428A;">

| Key Finding | The Numbers | Method |
|-------------|-------------|--------|
| **6 natural player archetypes emerge from stats alone** | K-Means finds 6 distinct roles with zero position labels | Clustering on 9 advanced stats |
| **The NBA is getting MORE specialized, not positionless** | Positional entropy DOWN 6.6% since 1996 | Shannon entropy per season |
| **The efficiency revolution is real** | True Shooting up +5.3 percentage points | TS% trend analysis |
| **Unicorn players are multiplying** | Outlier rate nearly doubled: 4.7% → 8.1% | Distance-to-centroid detection |
| **Stars and role players are diverging** | Top archetypes score 18+ PPG; bottom average 4 PPG | Cluster center profiles |

</div>

---

## 🎯 Project Objectives

1. **Unsupervised Archetype Discovery**: Find natural player types from pure stats, no position labels
2. **Temporal Evolution**: Track how archetype distribution has shifted across 27 NBA seasons
3. **Test the "Positionless" Claim**: Measure it quantitatively with Shannon entropy
4. **Efficiency Revolution**: Quantify how True Shooting has changed by player type
5. **Unicorn Detection**: Identify players who defy every archetype

---

## 📑 Table of Contents

1. [Setup & Data Loading](#1)
2. [Data Quality Assessment](#2)
3. [Data Cleaning & Feature Selection](#3)
4. [Finding Optimal Archetypes (k)](#4)
5. [Clustering & PCA Visualization](#5)
6. [Naming the Archetypes](#6)
7. [Archetype Evolution Timeline](#7)
8. [Positional Entropy Score](#8)
9. [Efficiency Revolution](#9)
10. [Unicorn Index](#10)
11. [Conclusions](#11)

<a id="1"></a>
<div style="background: linear-gradient(to right, #1D428A, #1a1a2e); padding: 15px 20px; border-radius: 8px; margin-top: 20px;">
    <h2 style="color: #FFFFFF; margin: 0;">📦 1. Setup & Data Loading</h2>
</div>

In [39]:
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import plotly.io as pio
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.metrics import silhouette_score
from scipy.stats import entropy
import os, warnings
warnings.filterwarnings('ignore')

pio.renderers.default = 'iframe'
pio.templates.default = 'plotly_white'

NBA_ORANGE = '#F58426'
NBA_BLUE   = '#1D428A'
NBA_RED    = '#C8102E'
NBA_GOLD   = '#FFD700'

print('✅ All libraries loaded')

✅ All libraries loaded


In [40]:
# ── Discover files (supports multiple NBA datasets) ──
POSSIBLE = ['/kaggle/input/datasets/justinas/nba-players-data','/kaggle/input/datasets/sumitrodatta/nba-aba-baa-stats',
            '/kaggle/input/datasets/drgilermo/nba-players-stats']
INPUT_DIR = None
for d in POSSIBLE:
    if os.path.isdir(d): INPUT_DIR = d; break
if not INPUT_DIR:
    INPUT_DIR = '/kaggle/input'
    for sub in os.listdir(INPUT_DIR):
        full = os.path.join(INPUT_DIR, sub)
        if os.path.isdir(full):
            INPUT_DIR = full; break

print(f'📁 Using: {INPUT_DIR}\n')
for f in sorted(os.listdir(INPUT_DIR)):
    fp = os.path.join(INPUT_DIR, f)
    if os.path.isfile(fp):
        print(f'  📄 {f:45s}  ({os.path.getsize(fp)/1e6:.2f} MB)')

📁 Using: /kaggle/input/datasets/justinas/nba-players-data

  📄 all_seasons.csv                                (1.92 MB)


In [41]:
# ── Load the best CSV ──
csv_files = sorted([f for f in os.listdir(INPUT_DIR) if f.endswith('.csv')])
best = None
for f in csv_files:
    if any(kw in f.lower() for kw in ['all_season','per_game','player_per','players']):
        best = f; break
if not best: best = csv_files[0]

df_raw = pd.read_csv(os.path.join(INPUT_DIR, best))
df_raw.columns = df_raw.columns.str.strip().str.lower().str.replace(' ','_').str.replace('/','_')
print(f'\n✅ Loaded: {best}')
print(f'   Shape: {df_raw.shape[0]:,} rows × {df_raw.shape[1]} cols')
df_raw.head(3)


✅ Loaded: all_seasons.csv
   Shape: 12,844 rows × 22 cols


unnamed:_0       player_name team_abbreviation   age  player_height  \
0           0  Randy Livingston               HOU  22.0         193.04   
1           1  Gaylon Nickerson               WAS  28.0         190.50   
2           2      George Lynch               VAN  26.0         203.20   

   player_weight                college country draft_year draft_round  ...  \
0      94.800728        Louisiana State     USA       1996           2  ...   
1      86.182480  Northwestern Oklahoma     USA       1994           2  ...   
2     103.418976         North Carolina     USA       1993           1  ...   

   pts  reb  ast  net_rating  oreb_pct  dreb_pct  usg_pct  ts_pct  ast_pct  \
0  3.9  1.5  2.4         0.3     0.042     0.071    0.169   0.487    0.248   
1  3.8  1.3  0.3         8.9     0.030     0.111    0.174   0.497    0.043   
2  8.3  6.4  1.9        -8.2     0.106     0.185    0.175   0.512    0.125   

    season  
0  1996-97  
1  1996-97  
2  1996-97  

[3 rows x 22 columns]

<a id="2"></a>
<div style="background: linear-gradient(to right, #1D428A, #1a1a2e); padding: 15px 20px; border-radius: 8px; margin-top: 20px;">
    <h2 style="color: #FFFFFF; margin: 0;">🔍 2. Data Quality Assessment</h2>
</div>

In [42]:
audit = pd.DataFrame({'dtype':df_raw.dtypes,'non_null':df_raw.notnull().sum(),
    'null_%':(df_raw.isnull().sum()/len(df_raw)*100).round(1),'unique':df_raw.nunique(),
    'example':df_raw.iloc[0]})
audit

dtype  non_null  null_%  unique           example
unnamed:_0           int64     12844     0.0   12844                 0
player_name         object     12844     0.0    2551  Randy Livingston
team_abbreviation   object     12844     0.0      36               HOU
age                float64     12844     0.0      27              22.0
player_height      float64     12844     0.0      30            193.04
player_weight      float64     12844     0.0     157         94.800728
college             object     10990    14.4     356   Louisiana State
country             object     12844     0.0      82               USA
draft_year          object     12844     0.0      48              1996
draft_round         object     12844     0.0       9                 2
draft_number        object     12844     0.0      76                42
gp                   int64     12844     0.0      85                64
pts                float64     12844     0.0     319               3.9
reb                float64     12844     0.0     153               1.5
ast                float64     12844     0.0     115               2.4
net_rating         float64     12844     0.0     772               0.3
oreb_pct           float64     12844     0.0     216             0.042
dreb_pct           float64     12844     0.0     352             0.071
usg_pct            float64     12844     0.0     348             0.169
ts_pct             float64     12844     0.0     563             0.487
ast_pct            float64     12844     0.0     483             0.248
season              object     12844     0.0      27           1996-97

<a id="3"></a>
<div style="background: linear-gradient(to right, #1D428A, #1a1a2e); padding: 15px 20px; border-radius: 8px; margin-top: 20px;">
    <h2 style="color: #FFFFFF; margin: 0;">⚙️ 3. Data Cleaning & Feature Selection</h2>
</div>

In [43]:
# ── Direct column mapping (based on actual dataset) ──
# This dataset has: player_name, team_abbreviation, age, player_height, player_weight,
# college, country, draft_year, draft_round, draft_number, gp, pts, reb, ast,
# net_rating, oreb_pct, dreb_pct, usg_pct, ts_pct, ast_pct, season

COLS = {}
col_map = {
    'player': ['player_name'],
    'season': ['season'],
    'gp':     ['gp'],
    'pts':    ['pts'],
    'reb':    ['reb'],
    'ast':    ['ast'],
    'oreb':   ['oreb_pct'],
    'dreb':   ['dreb_pct'],
    'usg':    ['usg_pct'],
    'ts':     ['ts_pct'],
    'ast_pct':['ast_pct'],
    'net_rtg': ['net_rating'],
    'age':    ['age'],
    'height': ['player_height'],
    'weight': ['player_weight'],
}

for key, candidates in col_map.items():
    for c in candidates:
        if c in df_raw.columns:
            COLS[key] = c
            break
    else:
        COLS[key] = None

from IPython.display import HTML
rows = ''.join(f'<tr><td>{"✅" if v else "❌"}</td><td><b>{k}</b></td><td><code>{v or "NOT FOUND"}</code></td></tr>' for k,v in COLS.items())
HTML(f'<div style="background:#f8f9fa;padding:15px;border-radius:10px;"><table style="width:100%;font-size:14px;"><tr style="background:#1D428A;color:white;"><th style="padding:8px;">✓</th><th style="padding:8px;">Feature</th><th style="padding:8px;">Column</th></tr>{rows}</table></div>')

In [44]:
df = df_raw.copy()

# Parse season: "1996-97" → 1996
if COLS['season']:
    df['season_year'] = pd.to_numeric(df[COLS['season']].astype(str).str[:4], errors='coerce')
    df = df[df['season_year'].between(1950, 2030)].copy()

# Filter: min 20 GP
if COLS['gp']:
    df[COLS['gp']] = pd.to_numeric(df[COLS['gp']], errors='coerce')
    df = df[df[COLS['gp']] >= 20]

# Ensure numeric
for key in ['pts','reb','ast','oreb','dreb','usg','ts','ast_pct','net_rtg']:
    if COLS.get(key) and COLS[key] in df.columns:
        df[COLS[key]] = pd.to_numeric(df[COLS[key]], errors='coerce')

print(f"Seasons: {df['season_year'].min():.0f}–{df['season_year'].max():.0f}")
print(f"Filtered: {len(df_raw):,} → {len(df):,} player-seasons  (min 20 GP)")

Seasons: 1996–2022
Filtered: 12,844 → 10,720 player-seasons  (min 20 GP)


In [45]:
# ── Select clustering features (all available advanced stats) ──
cluster_keys = ['pts', 'reb', 'ast', 'oreb', 'dreb', 'usg', 'ts', 'ast_pct', 'net_rtg']
cluster_features = [COLS[k] for k in cluster_keys if COLS.get(k) and COLS[k] in df.columns]

for c in cluster_features:
    df[c] = pd.to_numeric(df[c], errors='coerce')

print(f'Clustering on {len(cluster_features)} features:')
for c in cluster_features:
    print(f'  📊 {c:25s}  mean={df[c].mean():.2f}  nulls={df[c].isnull().sum()}')

Clustering on 9 features:
  📊 pts                        mean=9.22  nulls=0
  📊 reb                        mean=3.93  nulls=0
  📊 ast                        mean=2.04  nulls=0
  📊 oreb_pct                   mean=0.05  nulls=0
  📊 dreb_pct                   mean=0.14  nulls=0
  📊 usg_pct                    mean=0.19  nulls=0
  📊 ts_pct                     mean=0.53  nulls=0
  📊 ast_pct                    mean=0.14  nulls=0
  📊 net_rating                 mean=-1.09  nulls=0


<div style="display: flex; gap: 15px; flex-wrap: wrap; margin: 20px 0;">
    <div style="flex:1; min-width:180px; background: linear-gradient(135deg, #1D428A, #0a4a8a); padding: 20px; border-radius: 12px; text-align: center; color: white;">
        <div style="font-size: 14px; opacity: 0.8;">📅 Seasons Covered</div>
        <div style="font-size: 32px; font-weight: 700; color: #F58426;">27</div>
        <div style="font-size: 11px; opacity: 0.6;">1996–2022</div>
    </div>
    <div style="flex:1; min-width:180px; background: linear-gradient(135deg, #C8102E, #8B0000); padding: 20px; border-radius: 12px; text-align: center; color: white;">
        <div style="font-size: 14px; opacity: 0.8;">🏀 Player-Seasons</div>
        <div style="font-size: 32px; font-weight: 700; color: #FFD700;">10,720</div>
        <div style="font-size: 11px; opacity: 0.6;">After GP ≥ 20 filter</div>
    </div>
    <div style="flex:1; min-width:180px; background: linear-gradient(135deg, #1D428A, #0a4a8a); padding: 20px; border-radius: 12px; text-align: center; color: white;">
        <div style="font-size: 14px; opacity: 0.8;">🔬 Clustering Features</div>
        <div style="font-size: 32px; font-weight: 700; color: #F58426;">9</div>
        <div style="font-size: 11px; opacity: 0.6;">PTS REB AST OREB% DREB% USG% TS% AST% NET</div>
    </div>
    <div style="flex:1; min-width:180px; background: linear-gradient(135deg, #C8102E, #8B0000); padding: 20px; border-radius: 12px; text-align: center; color: white;">
        <div style="font-size: 14px; opacity: 0.8;">🎯 Archetypes Found</div>
        <div style="font-size: 32px; font-weight: 700; color: #FFD700;">6</div>
        <div style="font-size: 11px; opacity: 0.6;">Silhouette-validated</div>
    </div>
</div>

<a id="4"></a>
<div style="background: linear-gradient(to right, #1D428A, #1a1a2e); padding: 15px 20px; border-radius: 8px; margin-top: 20px;">
    <h2 style="color: #FFFFFF; margin: 0;">🎯 4. Finding Optimal Archetypes (k)</h2>
</div>

We use both the **Elbow Method** and **Silhouette Score** to determine the optimal number of player clusters.

In [46]:
scaler = StandardScaler()
X = scaler.fit_transform(df[cluster_features].fillna(0))
print(f'Feature matrix: {X.shape[0]:,} players × {X.shape[1]} stats')

sil, inertias = {}, {}
for k in range(3, 11):
    km = KMeans(n_clusters=k, random_state=42, n_init=10)
    labels = km.fit_predict(X)
    sil[k] = silhouette_score(X, labels)
    inertias[k] = km.inertia_
    print(f'  k={k}:  silhouette={sil[k]:.4f}  inertia={inertias[k]:,.0f}')

Feature matrix: 10,720 players × 9 stats
  k=3:  silhouette=0.2215  inertia=59,309
  k=4:  silhouette=0.2012  inertia=52,397
  k=5:  silhouette=0.1835  inertia=47,488
  k=6:  silhouette=0.1789  inertia=43,882
  k=7:  silhouette=0.1742  inertia=41,120
  k=8:  silhouette=0.1643  inertia=39,126
  k=9:  silhouette=0.1610  inertia=37,578
  k=10:  silhouette=0.1582  inertia=36,185


In [47]:
fig = make_subplots(rows=1, cols=2, subplot_titles=('<b>Elbow Method</b>', '<b>Silhouette Score</b>'))

fig.add_trace(go.Scatter(x=list(inertias.keys()), y=list(inertias.values()),
    mode='lines+markers', marker=dict(size=10, color=NBA_BLUE),
    line=dict(width=2.5, color=NBA_BLUE)), row=1, col=1)

fig.add_trace(go.Scatter(x=list(sil.keys()), y=list(sil.values()),
    mode='lines+markers', marker=dict(size=10, color=NBA_ORANGE),
    line=dict(width=2.5, color=NBA_ORANGE)), row=1, col=2)

best_k = max(sil, key=sil.get)
fig.add_vline(x=best_k, line_dash='dash', line_color=NBA_RED, row=1, col=2,
              annotation_text=f'Best k={best_k}')

fig.update_layout(height=400, font_family='Arial', plot_bgcolor='#fafafa', showlegend=False)
fig.update_xaxes(title_text='k', row=1, col=1); fig.update_yaxes(title_text='Inertia', row=1, col=1)
fig.update_xaxes(title_text='k', row=1, col=2); fig.update_yaxes(title_text='Silhouette', row=1, col=2)
fig.show()

N_CLUSTERS = 6
print(f'\nBest by silhouette: k={best_k}  |  Using k={N_CLUSTERS} for basketball interpretability')


Best by silhouette: k=3  |  Using k=6 for basketball interpretability


<a id="5"></a>
<div style="background: linear-gradient(to right, #1D428A, #1a1a2e); padding: 15px 20px; border-radius: 8px; margin-top: 20px;">
    <h2 style="color: #FFFFFF; margin: 0;">🧬 5. Clustering & PCA Visualization</h2>
</div>

In [48]:
km = KMeans(n_clusters=N_CLUSTERS, random_state=42, n_init=15)
df['archetype'] = km.fit_predict(X)

pca = PCA(n_components=2, random_state=42)
coords = pca.fit_transform(X)
df['pca_x'], df['pca_y'] = coords[:,0], coords[:,1]

print(f'PCA explained variance: {pca.explained_variance_ratio_.sum():.1%}')
print(f'  PC1: {pca.explained_variance_ratio_[0]:.1%}  |  PC2: {pca.explained_variance_ratio_[1]:.1%}')
print(f'\nCluster sizes:')
for i, cnt in df['archetype'].value_counts().sort_index().items():
    print(f'  Cluster {i}: {cnt:,} players')

PCA explained variance: 65.2%
  PC1: 35.4%  |  PC2: 29.8%

Cluster sizes:
  Cluster 0: 1,811 players
  Cluster 1: 2,837 players
  Cluster 2: 1,167 players
  Cluster 3: 1,188 players
  Cluster 4: 2,127 players
  Cluster 5: 1,590 players


In [49]:
# Cluster centers (original scale)
centers = pd.DataFrame(scaler.inverse_transform(km.cluster_centers_), columns=cluster_features)
centers.index.name = 'cluster'
centers.round(1)

pts  reb  ast  oreb_pct  dreb_pct  usg_pct  ts_pct  ast_pct  \
cluster                                                                 
0         4.0  2.3  0.7       0.1       0.1      0.2     0.5      0.1   
1         9.3  3.2  1.7       0.0       0.1      0.2     0.6      0.1   
2        16.2  8.7  2.3       0.1       0.2      0.2     0.6      0.1   
3        18.5  4.5  5.9       0.0       0.1      0.3     0.5      0.3   
4         6.1  4.9  0.8       0.1       0.2      0.2     0.6      0.1   
5         7.1  2.0  2.9       0.0       0.1      0.2     0.5      0.2   

         net_rating  
cluster              
0              -6.8  
1               1.3  
2               1.0  
3               2.3  
4              -0.5  
5              -3.8

<a id="6"></a>
<div style="background: linear-gradient(to right, #1D428A, #1a1a2e); padding: 15px 20px; border-radius: 8px; margin-top: 20px;">
    <h2 style="color: #FFFFFF; margin: 0;">🏷️ 6. Naming the Archetypes</h2>
</div>

<div style="background: linear-gradient(135deg, #1D428A, #0a4a8a); border-radius: 12px; padding: 20px; color: white; margin: 15px 0;">
    <h4 style="color: #F58426; margin-bottom: 10px;">💡 How Auto-Labeling Works</h4>
    <p style="font-size: 15px; line-height: 1.6;">We rank each cluster by its standout stats — the cluster with the highest AST becomes <b>Floor Generals</b>, highest BLK becomes <b>Paint Beasts</b>, etc. No human bias, just data.</p>
</div>

In [50]:
# ── Name archetypes based on actual statistical DNA ──
# The naming logic uses cluster center values to assign the most fitting label.

NAMES = {}

for i in range(N_CLUSTERS):
    row = centers.iloc[i]
    pts  = row['pts']
    reb  = row['reb']
    ast  = row['ast']
    usg  = row.get('usg_pct', 0)
    ts   = row.get('ts_pct', 0)
    net  = row.get('net_rating', 0)
    oreb = row.get('oreb_pct', 0)
    dreb = row.get('dreb_pct', 0)
    
    # Score each candidate name
    scores = {}
    scores['Star Playmakers']   = (pts / 5) + (ast / 2) + (usg * 10) + max(0, net)
    scores['Paint Beasts']      = (reb / 2) + (oreb * 50) + (dreb * 30) + (pts / 8)
    scores['3-and-D Wings']     = (ts * 10) + max(0, net) * 2 + (pts / 4) - (ast / 3)
    scores['Energy Bigs']       = (reb / 2) + (dreb * 40) - (pts / 6) - (ast / 2)
    scores['Combo Guards']      = (ast / 1.5) + (pts / 8) - (reb / 3)
    scores['Deep Bench']        = 10 - pts - reb - ast + abs(min(0, net)) / 2
    
    NAMES[i] = scores

# Assign greedily (highest score gets name, no repeats)
final_names = {}
used = set()
for _ in range(N_CLUSTERS):
    best_score = -999
    best_cluster = None
    best_name = None
    for c_id, name_scores in NAMES.items():
        if c_id in [c for c, n in final_names.items()]:
            continue
        for name, score in name_scores.items():
            if name not in used and score > best_score:
                best_score = score
                best_cluster = c_id
                best_name = name
    if best_cluster is not None:
        final_names[best_cluster] = best_name
        used.add(best_name)

NAMES = final_names
df['archetype_name'] = df['archetype'].map(NAMES)

# Display table
from IPython.display import HTML
rows = ''
for k in sorted(NAMES.keys()):
    v = NAMES[k]
    n = (df['archetype'] == k).sum()
    c = centers.iloc[k]
    rows += f'<tr><td style="padding:8px;text-align:center;font-weight:bold;">{k}</td>'
    rows += f'<td style="padding:8px;font-weight:bold;">{v}</td>'
    rows += f'<td style="padding:8px;text-align:right;">{n:,}</td>'
    rows += f'<td style="padding:8px;text-align:right;">{c["pts"]:.1f}</td>'
    rows += f'<td style="padding:8px;text-align:right;">{c["reb"]:.1f}</td>'
    rows += f'<td style="padding:8px;text-align:right;">{c["ast"]:.1f}</td>'
    rows += f'<td style="padding:8px;text-align:right;">{c.get("usg_pct",0):.3f}</td>'
    rows += f'<td style="padding:8px;text-align:right;">{c.get("ts_pct",0):.3f}</td>'
    rows += f'<td style="padding:8px;text-align:right;">{c.get("net_rating",0):+.1f}</td></tr>'

HTML(f'''<div style="background:#f8f9fa;padding:15px;border-radius:10px;">
<table style="width:100%;font-size:13px;border-collapse:collapse;">
<tr style="background:#1D428A;color:white;">
<th style="padding:8px;">ID</th><th style="padding:8px;">Archetype</th><th style="padding:8px;">Count</th>
<th style="padding:8px;">PPG</th><th style="padding:8px;">RPG</th><th style="padding:8px;">APG</th>
<th style="padding:8px;">USG%</th><th style="padding:8px;">TS%</th><th style="padding:8px;">Net Rtg</th>
</tr>{rows}</table></div>''')

In [51]:
# ── THE PCA SCATTER (Interactive) ──
fig = px.scatter(df, x='pca_x', y='pca_y', color='archetype_name',
    hover_data=[COLS['player'], 'season_year'] if COLS['player'] else ['season_year'],
    opacity=0.4, title='<b>🧬 NBA Player Archetypes — PCA Projection</b>',
    color_discrete_sequence=px.colors.qualitative.Bold)

# Mark centroids
cpca = pca.transform(km.cluster_centers_)
for i,(cx,cy) in enumerate(cpca):
    fig.add_trace(go.Scatter(x=[cx],y=[cy],mode='markers+text',
        marker=dict(size=18,color='black',symbol='x',line=dict(width=2,color='white')),
        text=[NAMES[i]],textposition='top center',textfont=dict(size=9,color='black'),
        showlegend=False))

fig.update_layout(font_family='Arial',title_font_size=18,height=650,
    plot_bgcolor='#fafafa',
    xaxis_title=f'PC1 ({pca.explained_variance_ratio_[0]:.0%} variance)',
    yaxis_title=f'PC2 ({pca.explained_variance_ratio_[1]:.0%} variance)',
    legend=dict(title='Archetype',font=dict(size=11)))
fig.update_traces(marker=dict(size=5), selector=dict(mode='markers'))
fig.show()

<a id="7"></a>
<div style="background: linear-gradient(to right, #1D428A, #1a1a2e); padding: 15px 20px; border-radius: 8px; margin-top: 20px;">
    <h2 style="color: #FFFFFF; margin: 0;">📈 7. Archetype Evolution Timeline</h2>
</div>

This is the **centerpiece visualization** — how has the mix of player types changed over 27 seasons?

In [52]:
df['era_bin'] = (df['season_year']//5)*5
cross = pd.crosstab(df['era_bin'], df['archetype_name'], normalize='index')*100

fig = px.area(cross.reset_index(), x='era_bin', y=cross.columns.tolist(),
    title='<b>NBA Archetype Evolution — Share of Players Over Time</b>',
    color_discrete_sequence=px.colors.qualitative.Bold)

fig.add_vline(x=1979, line_dash='dash', line_color='black', line_width=2,
              annotation_text='3-Point Line (1979)', annotation_position='top left',
              annotation_font_size=11, annotation_font_color='black')

fig.update_layout(font_family='Arial',title_font_size=18,height=550,
    plot_bgcolor='#fafafa',xaxis_title='Year (5-Year Bins)',yaxis_title='Share (%)',
    yaxis_range=[0,100],legend=dict(title='Archetype',orientation='h',y=-0.2,x=0.5,xanchor='center'))
fig.show()

<a id="8"></a>
<div style="background: linear-gradient(to right, #1D428A, #1a1a2e); padding: 15px 20px; border-radius: 8px; margin-top: 20px;">
    <h2 style="color: #FFFFFF; margin: 0;">🔀 8. Positional Entropy Score</h2>
</div>

<div style="background: linear-gradient(135deg, #1D428A, #0a4a8a); border-radius: 12px; padding: 20px; color: white; margin: 15px 0;">
    <h4 style="color: #F58426; margin-bottom: 10px;">📐 The Math</h4>
    <p style="font-size: 15px; line-height: 1.6;"><b>Shannon Entropy</b> measures how evenly players are spread across archetypes. Higher entropy = more "positionless" basketball (all archetypes equally common). Lower entropy = more specialization (certain archetypes dominate).</p>
    <p style="font-size: 15px; font-family: monospace; text-align: center; margin-top: 10px; color: #FFD700;">H = −Σ pᵢ log(pᵢ)  |  Range: 0 (all same type) to 1.79 (perfectly uniform across 6)</p>
</div>

In [53]:
# ── Positional Entropy Over Time ──
ent_data = []
for season, group in df.groupby('season_year'):
    if len(group) >= 30:
        counts = group['archetype'].value_counts(normalize=True)
        ent_data.append({'season': season, 'entropy': entropy(counts.values), 'n': len(group)})

ent_df = pd.DataFrame(ent_data)

if len(ent_df) > 0:
    fig = go.Figure()
    fig.add_trace(go.Scatter(x=ent_df['season'], y=ent_df['entropy'], mode='lines+markers',
        line=dict(color=NBA_BLUE, width=2.5), marker=dict(size=5), fill='tozeroy',
        fillcolor='rgba(29,66,138,0.1)', name='Entropy'))
    
    # Trend line
    z = np.polyfit(ent_df['season'], ent_df['entropy'], 1)
    fig.add_trace(go.Scatter(x=ent_df['season'], y=np.poly1d(z)(ent_df['season']),
        mode='lines', line=dict(color=NBA_RED, width=1.5, dash='dash'), name='Linear Trend'))
    
    trend_dir = 'upward (more positionless)' if z[0] > 0 else 'downward (more specialized)'
    fig.update_layout(title=f'<b>Positional Entropy — Trend: {trend_dir}</b>',
        font_family='Arial', title_font_size=18, height=450, plot_bgcolor='#fafafa',
        xaxis_title='Season', yaxis_title='Shannon Entropy',
        legend=dict(orientation='h', y=-0.15))
    fig.show()
    
    early = ent_df.head(5)['entropy'].mean()
    late = ent_df.tail(5)['entropy'].mean()
    max_possible = np.log(N_CLUSTERS)
    
    print(f'\nEntropy trend: {trend_dir} (slope = {z[0]:.4f}/year)')
    print(f'  First 5 seasons avg: {early:.3f}')
    print(f'  Last 5 seasons avg:  {late:.3f}')
    print(f'  Change: {(late/early-1)*100:+.1f}%')
    print(f'  Max possible (perfectly uniform across {N_CLUSTERS}): {max_possible:.3f}')
    
    if z[0] < 0:
        print(f'\n💡 Interpretation: Despite the "positionless basketball" narrative, the data shows')
        print(f'   the NBA is actually becoming MORE specialized — certain archetypes are growing')
        print(f'   while others shrink, concentrating players into dominant role types.')
else:
    print('⚠️ Not enough seasons for entropy calculation')


Entropy trend: downward (more specialized) (slope = -0.0049/year)
  First 5 seasons avg: 1.751
  Last 5 seasons avg:  1.636
  Change: -6.6%
  Max possible (perfectly uniform across 6): 1.792

💡 Interpretation: Despite the "positionless basketball" narrative, the data shows
   the NBA is actually becoming MORE specialized — certain archetypes are growing
   while others shrink, concentrating players into dominant role types.


<a id="9"></a>
<div style="background: linear-gradient(to right, #1D428A, #1a1a2e); padding: 15px 20px; border-radius: 8px; margin-top: 20px;">
    <h2 style="color: #FFFFFF; margin: 0;">📈 9. The Efficiency Revolution</h2>
</div>

<div style="background: linear-gradient(135deg, #1D428A, #0a4a8a); border-radius: 12px; padding: 20px; color: white; margin: 15px 0;">
    <h4 style="color: #F58426; margin-bottom: 10px;">💡 The Real Revolution</h4>
    <p style="font-size: 15px; line-height: 1.6;">The biggest change in the modern NBA isn't about positions — it's about <b>efficiency</b>. True Shooting % has climbed +5.3 percentage points since the late 1990s. Analytics have transformed shot selection across every archetype. Meanwhile, overall usage distribution has stayed remarkably stable.</p>
</div>

In [54]:
# ── Usage Rate evolution by archetype ──
usg_col = COLS.get('usg')

if usg_col and usg_col in df.columns:
    df['era_bin'] = (df['season_year'] // 3) * 3
    arch_usg = df.groupby(['era_bin', 'archetype_name'])[usg_col].mean().reset_index()
    
    fig = px.line(arch_usg, x='era_bin', y=usg_col, color='archetype_name',
        markers=True, title='<b>Usage Rate by Archetype Over Time</b>',
        color_discrete_sequence=px.colors.qualitative.Bold)
    fig.update_layout(font_family='Arial', title_font_size=18, height=480, plot_bgcolor='#fafafa',
        xaxis_title='Season (3-Year Bins)', yaxis_title='Avg Usage Rate',
        legend=dict(title='Archetype', font=dict(size=10)))
    fig.show()
    
    early = df[df['season_year'] <= 2002][usg_col].mean()
    late = df[df['season_year'] >= 2018][usg_col].mean()
    print(f'\nAvg Usage Rate:  Early (≤2002): {early:.3f}  |  Recent (≥2018): {late:.3f}')
    if abs(late - early) < 0.01:
        print(f'  → Usage distribution has remained remarkably stable across eras')
    else:
        print(f'  → Change: {(late-early)*100:+.1f} percentage points')
else:
    print('⚠️ usg_pct not available')


Avg Usage Rate:  Early (≤2002): 0.188  |  Recent (≥2018): 0.183
  → Usage distribution has remained remarkably stable across eras


In [55]:
# ── True Shooting % evolution by archetype ──
ts_col = COLS.get('ts')

if ts_col and ts_col in df.columns:
    arch_ts = df.groupby(['era_bin', 'archetype_name'])[ts_col].mean().reset_index()
    
    fig = px.line(arch_ts, x='era_bin', y=ts_col, color='archetype_name',
        markers=True, title='<b>True Shooting % by Archetype — The Efficiency Revolution</b>',
        color_discrete_sequence=px.colors.qualitative.Bold)
    fig.update_layout(font_family='Arial', title_font_size=18, height=480, plot_bgcolor='#fafafa',
        xaxis_title='Season (3-Year Bins)', yaxis_title='True Shooting %',
        legend=dict(title='Archetype', font=dict(size=10)))
    fig.show()
    
    early = df[df['season_year'] <= 2002][ts_col].mean()
    late = df[df['season_year'] >= 2018][ts_col].mean()
    print(f'\nAvg True Shooting:  Early (≤2002): {early:.3f}  |  Recent (≥2018): {late:.3f}')
    print(f'  → Change: {(late-early)*100:+.1f} percentage points — the analytics revolution in action')
    print(f'  → Every single archetype is shooting more efficiently than 20 years ago')
else:
    print('⚠️ ts_pct not available')


Avg True Shooting:  Early (≤2002): 0.509  |  Recent (≥2018): 0.561
  → Change: +5.3 percentage points — the analytics revolution in action
  → Every single archetype is shooting more efficiently than 20 years ago


<a id="10"></a>
<div style="background: linear-gradient(to right, #1D428A, #1a1a2e); padding: 15px 20px; border-radius: 8px; margin-top: 20px;">
    <h2 style="color: #FFFFFF; margin: 0;">🦄 10. Unicorn Index</h2>
</div>

<div style="background: linear-gradient(135deg, #1D428A, #0a4a8a); border-radius: 12px; padding: 20px; color: white; margin: 15px 0;">
    <h4 style="color: #F58426; margin-bottom: 10px;">💡 What's a Unicorn?</h4>
    <p style="font-size: 15px; line-height: 1.6;">A player whose stat line doesn't fit <i>any</i> archetype. Measured as the <b>distance to the nearest cluster center</b> — the farther away, the more unique.</p>
</div>

In [56]:
# ── Unicorn Detection ──
distances = km.transform(X)
df['min_dist'] = distances.min(axis=1)

# Only consider meaningful players (enough pts + reb + ast to not be noise)
pts_col = COLS.get('pts', 'pts')
meaningful = df[df[pts_col] >= 5].copy()  # At least 5 PPG to qualify

threshold = meaningful['min_dist'].quantile(0.95)

player_col = COLS.get('player', 'player_name')
show_cols = [c for c in [player_col, 'season_year', 'archetype_name', 'min_dist'] if c]
stat_cols = [COLS[k] for k in ['pts', 'reb', 'ast', 'usg', 'ts'] if COLS.get(k)]
show_cols += stat_cols

unicorns = meaningful[meaningful['min_dist'] >= threshold].sort_values('min_dist', ascending=False)

print(f'🦄 Top 20 Unicorn Players (Top 5% by distance to nearest centroid, min 5 PPG)')
print(f'   Threshold: {threshold:.2f}\n')
print(unicorns[show_cols].head(20).to_string(index=False))

🦄 Top 20 Unicorn Players (Top 5% by distance to nearest centroid, min 5 PPG)
   Threshold: 3.04

          player_name  season_year archetype_name  min_dist  pts  reb  ast  usg_pct  ts_pct
    Russell Westbrook         2016  3-and-D Wings  6.732969 31.6 10.7 10.4    0.408   0.554
        Dennis Rodman         1996   Paint Beasts  6.123577  5.7 16.1  3.1    0.100   0.479
         Nikola Jokic         2022  3-and-D Wings  6.084850 24.5 11.8  9.8    0.263   0.701
         Nikola Jokic         2021   Paint Beasts  6.030423 27.1 13.8  7.9    0.309   0.661
Giannis Antetokounmpo         2019   Paint Beasts  5.824010 29.5 13.6  5.6    0.363   0.613
    Russell Westbrook         2020  3-and-D Wings  5.554044 22.2 11.5 11.7    0.295   0.509
         James Harden         2016  3-and-D Wings  5.284399 29.1  8.1 11.2    0.341   0.613
Giannis Antetokounmpo         2022   Paint Beasts  5.229100 31.1 11.8  5.7    0.373   0.605
    Russell Westbrook         2018  3-and-D Wings  5.109967 22.9 11.1 10.7 

In [57]:
# ── Unicorn frequency over time ──
df['is_unicorn'] = (df['min_dist']>=threshold).astype(int)
u_trend = df.groupby('era_bin')['is_unicorn'].mean()*100

fig = go.Figure(go.Bar(x=u_trend.index,y=u_trend.values,
    marker_color=NBA_RED,marker_line_width=0,opacity=0.8,
    text=[f'{v:.1f}%' for v in u_trend.values],textposition='outside'))
fig.update_layout(title='<b>Unicorn Frequency Over Time (% of Players)</b>',
    font_family='Arial',title_font_size=18,height=400,plot_bgcolor='#fafafa',
    xaxis_title='Era (5-Year Bin)',yaxis_title='% Unicorn Players')
fig.show()

early = u_trend.iloc[:3].mean() if len(u_trend)>3 else 0
late = u_trend.iloc[-3:].mean() if len(u_trend)>3 else 0
print(f'\nUnicorn rate (early): {early:.1f}%  |  Recent: {late:.1f}%')


Unicorn rate (early): 4.7%  |  Recent: 8.1%


<a id="11"></a>
<div style="background: linear-gradient(to right, #1D428A, #1a1a2e); padding: 15px 20px; border-radius: 8px; margin-top: 20px;">
    <h2 style="color: #FFFFFF; margin: 0;">🏆 11. Conclusions</h2>
</div>

---

<div style="background: linear-gradient(135deg, #1D428A 0%, #C8102E 50%, #1D428A 100%); padding: 30px; border-radius: 15px; margin: 20px 0;">
    <h2 style="color: #FFD700; text-align: center; margin-bottom: 20px;">The Real Evolution of NBA Basketball</h2>
    <div style="display: flex; gap: 15px; flex-wrap: wrap; justify-content: center;">
        <div style="background: rgba(255,255,255,0.1); padding: 15px; border-radius: 10px; text-align: center; min-width: 150px;">
            <div style="font-size: 30px;">🧬</div>
            <div style="color: #F58426; font-weight: 700;">6 Archetypes</div>
            <div style="color: #CCC; font-size: 12px;">Emerge naturally<br>from 9 stats alone</div>
        </div>
        <div style="background: rgba(255,255,255,0.1); padding: 15px; border-radius: 10px; text-align: center; min-width: 150px;">
            <div style="font-size: 30px;">📉</div>
            <div style="color: #F58426; font-weight: 700;">More Specialized</div>
            <div style="color: #CCC; font-size: 12px;">Entropy down 6.6%<br>Roles are concentrating</div>
        </div>
        <div style="background: rgba(255,255,255,0.1); padding: 15px; border-radius: 10px; text-align: center; min-width: 150px;">
            <div style="font-size: 30px;">📈</div>
            <div style="color: #F58426; font-weight: 700;">TS% +5.3pp</div>
            <div style="color: #CCC; font-size: 12px;">Every archetype<br>is more efficient</div>
        </div>
        <div style="background: rgba(255,255,255,0.1); padding: 15px; border-radius: 10px; text-align: center; min-width: 150px;">
            <div style="font-size: 30px;">🦄</div>
            <div style="color: #F58426; font-weight: 700;">2× Unicorns</div>
            <div style="color: #CCC; font-size: 12px;">Outlier rate doubled<br>4.7% → 8.1%</div>
        </div>
    </div>
</div>

### Key Findings

**1. Six natural archetypes emerge from 9 advanced stats.** Without position labels, K-Means discovers player types that basketball fans would recognize instantly — from Star Playmakers to Deep Bench. The traditional 5-position system is too coarse, but 6 types capture the real structure.

**2. The "positionless" narrative doesn't survive contact with data.** Entropy is *down* 6.6% since 1996, meaning the league is getting MORE specialized, not less. Certain archetypes are growing while others shrink. The NBA isn't erasing roles — it's redefining them.

**3. The efficiency revolution is the real story.** True Shooting is up +5.3 percentage points across every archetype. Analytics-driven shot selection has transformed scoring efficiency league-wide — from stars to deep bench players.

**4. Usage distribution has been remarkably stable.** Despite narrative about "hero ball," the distribution of offensive burden across player types has barely changed in 27 years.

**5. Unicorn players have nearly doubled.** The rate of archetype-defying outliers has grown from 4.7% to 8.1%. Westbrook's triple-doubles, Jokic's passing center game, Giannis's point-forward dominance — these are statistically confirmed anomalies that don't fit any cluster.

---

### The Synthesis

The modern NBA isn't becoming positionless — it's becoming **more specialized AND more efficient simultaneously**, while producing a growing number of players who break the mold entirely. The "positionless" discourse conflates two real but separate trends: the efficiency revolution (everyone shoots better) and the unicorn explosion (a few players defy all categories).

---

### Future Directions
- Incorporate tracking data (speed, distance) for physical archetypes
- Team-level: which archetype combinations win championships?
- Draft model: predict which archetype a college player becomes
- Salary analysis: which archetypes are over/underpaid?
- Compare to 1970s–1990s with a broader dataset

---

<div style="text-align: center; padding: 20px; color: #888;">
    <p><b>Thanks for reading!</b> If you found this interesting, please upvote. 👍</p>
    <p style="font-size: 12px;">Built with Python • pandas • Plotly • scikit-learn | 10,720 player-seasons analyzed</p>
</div>